flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/KafkaDynamicSource.java

```diff
@@ -105,24 +122,37 @@
 
 	public KafkaDynamicSource(
 			DataType physicalDataType,
+			ChangelogMode changelogMode,
+			@Nullable DecodingFormat<DeserializationSchema<RowData>> keyDecodingFormat,
+			DecodingFormat<DeserializationSchema<RowData>> valueDecodingFormat,
+			int[] keyProjection,
+			int[] valueProjection,
+			@Nullable String keyPrefix,
 			@Nullable List<String> topics,
 			@Nullable Pattern topicPattern,
 			Properties properties,
-			DecodingFormat<DeserializationSchema<RowData>> decodingFormat,
 			StartupMode startupMode,
 			Map<KafkaTopicPartition, Long> specificStartupOffsets,
 			long startupTimestampMillis) {
+		// Format attributes
 		this.physicalDataType = Preconditions.checkNotNull(physicalDataType, "Physical data type must not be null.");
+		this.keyDecodingFormat = keyDecodingFormat;
+		this.valueDecodingFormat = Preconditions.checkNotNull(
+				valueDecodingFormat, "Value decoding format must not be null.");
+		this.keyProjection = Preconditions.checkNotNull(keyProjection, "Key projection must not be null.");
+		this.valueProjection = Preconditions.checkNotNull(valueProjection, "Value projection must not be null.");
+		this.keyPrefix = keyPrefix;
+		// Mutable attributes
 		this.producedDataType = physicalDataType;
 		this.metadataKeys = Collections.emptyList();
+		this.changelogMode = changelogMode;
```

Comment:
From my point of view, privoding the `ChangelogMode` by invokers is not safe. 
A more elegant way is providing a wrapper of `DecodingFormat` and `EncodingFormat` by the table factory, e.g. the upsert kafka factory should provide a wrapper which returns upsert changelog mode for `DecodingFormat#getChangelogMode()`. Then there is nothing to change in the `KafkaDynamicSource` and `KafkaDynamicSink`.

I think you need the wrapper anyway, because you need to change the RowKind from `INSERT` to `UPDATE_AFTER`.